Web Driver: add support for navigation events

AnujPanta1 opened the pull request:
#### a9cb9e78eda3580235bbedcb8584a983046f7494
<pre>
WebDriver BiDi: support for navigation-related events in `browsingContext`
<a href="https://bugs.webkit.org/show_bug.cgi?id=297278">https://bugs.webkit.org/show_bug.cgi?id=297278</a>
<a href="https://rdar.apple.com/137019783">rdar://137019783</a>

Reviewed by NOBODY (OOPS!).

Navigation events are now hooked into WebKit&apos;s existing page load lifecycle callbacks:
didCommitLoadForFrame() for committed navigation,
didFailProvisionalLoadForFrameShared() and didFailLoadForFrame() for failed events,
didCancelClientRedirectForFrame() for aborted navigation,
didSameDocumentNavigationForFrame() variants for fragment navigation.

The implementation extends the initial navigationStarted foundation. A new helper
convertNavigationIDToString() added.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::convertNavigationIDToString):
(WebKit::WebAutomationSession::navigationStartedForFrame):
(WebKit::WebAutomationSession::navigationCommittedForFrame):
(WebKit::WebAutomationSession::navigationFailedForFrame):
(WebKit::WebAutomationSession::navigationAbortedForFrame):
(WebKit::WebAutomationSession::fragmentNavigatedForFrame):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCancelClientRedirectForFrame):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didFailLoadForFrame):
(WebKit::WebPageProxy::didSameDocumentNavigationForFrame):
(WebKit::WebPageProxy::didSameDocumentNavigationForFrameViaJS):
</pre>

jesxilin commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6822,6 +6822,14 @@ void WebPageProxy::didStartProvisionalLoadForFrameShared(Ref<WebProcessProxy>&&
+    if (m_controlledByAutomation) {
+        if (RefPtr automationSession = process->processPool().automationSession()) {
+            automationSession->navigationStartedForFrame(*frame, url.string(),
+                navigationID ? std::optional<String>(String::number(navigationID->toUInt64())) : std::nullopt,
```
Nit suggestion: To me, this line is a bit confusing, but it seems to be a null check on `navigationID`. Perhaps it's possible to simplify this somehow... since `navigationID` can either be a string, empty string, or null, I don't think `std::optional<String>` is needed.

burg replied:
I'd prefer that for hooks like this, we pass WebKit's `navigationID` onward using its original type. Any WebDriver-specific massaging of the type or value should be in `WebAutomationSession`, `WebDriverBidiProcessor`, or command implementations in `Bidi*Agent`.

jesxilin commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -951,6 +951,12 @@ void WebAutomationSession::didCreatePage(WebPageProxy& page)
 }
+
+void WebAutomationSession::navigationStartedForFrame(const WebFrameProxy& frame, const String& url, std::optional<String> navigationID, double timestamp)
+{
```
Pertaining my other nit comment, we have String methods like `nullString()` and `emptyString()` that can be used inside the function to check if you choose to not use `std::optional<String>`.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -153,6 +165,17 @@
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
+                { "name": "timestamp", "type": "number" },
```
The spec says:

> Let timestamp be a [time value](https://tc39.es/ecma262/#sec-time-values-and-time-range) representing the current date and time in UTC.

On a quick read I'm not sure if this is milliseconds, seconds, or what. Please specify what the number value means in the `timestamp` parameter description.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -153,6 +165,17 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#event-browsingContext-navigationStarted",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
```
I read the spec and while it definitely specifies a nullable `navigation`, I couldn't intuit where this situation occurs for this specific event. So I read deeper and it says in the intro to the `browsingContext` section that

> id
The [navigation id](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigation-id) for the navigation, or null when the navigation is canceled before making progress.

Looking into this more, in `WebPageProxy` it uses 

`std::optional<WebCore::NavigationIdentifier>`, which is an optional `ObjectIdentifier<NavigationIdentifierType, uint64_t>`.

However, the WebDriver Bidi spec references the navigation ID defined [elsewhere] as UUID(https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigation-id):

> A navigation ID is a UUID string generated during navigation. It is used to interface with the WebDriver BiDi specification as well as to track the [ongoing navigation](https://html.spec.whatwg.org/multipage/browsing-the-web.html#ongoing-navigation). [[WEBDRIVERBIDI]](https://html.spec.whatwg.org/multipage/references.html#refsWEBDRIVERBIDI)
`

So putting these two findings together, we get `std::optional<uint64_t>` from WebCore and WebDriver Bidi reports out optional UUID strings. Let's come up with a mapping from uint64 to UUID-maybe `HashMap<uint64_t, UUID>` in WebAutomationSession.

Separately, we should be curious about WebDriver Bidi's notion of navigation "progress" and whether that differs from WebKit's navigation system. As-is, this hook will be called whether or not we make any "network progress" by receiving data for the main resource. For now it's best to keep going and land this PR so that we can unblock more testing, but keep this in mind when we start looking into test failures in this area.

AnujPanta1 replied:
Yes I'll note this down on the tracking sheet's notes to remember. As for the difference in the navigation progress I yes the tests are the best way to see if the difference is drastic

burg reviewed (commented):
I'll approve this once we get all the bots green.

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6822,6 +6822,14 @@ void WebPageProxy::didStartProvisionalLoadForFrameShared(Ref<WebProcessProxy>&&
     frame->setUnreachableURL(unreachableURL);
     frame->didStartProvisionalLoad(WTFMove(url));
 
+    if (m_controlledByAutomation) {
```
You added the call to `navigationStartedForFrame()` but it's missing the guard `#if ENABLE(WEBDRIVER_BIDI)`, which is why tvOS build is failing.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -951,6 +951,18 @@ void WebAutomationSession::didCreatePage(WebPageProxy& page)
+void WebAutomationSession::navigationStartedForFrame(const WebFrameProxy& frame, const String& url, std::optional<WebCore::NavigationIdentifier> navigationID, double timestamp)
+{
+    String browsingContextHandle = handleForWebFrameProxy(frame);
+    String navigationIDString;
```
We have a difference between `optional` and `nullable` in the protocol generator, but for event parameters I think we need to do the encoding of null string ourselves. I believe that in the case where navigationID is nullopt, this should be nullString(), right?

AnujPanta1 replied:
Yes I'll make it the default value for string (a 1 line else statement causes the style checker to get mad)

AnujPanta1 committed:
Web Driver: add support for navigation events
https://bugs.webkit.org/show_bug.cgi?id=296759
rdar://157234247

Reviewed by BJ Burg.

When a frame begins navigation, we capture the event at
WebPageProxy::didStartProvisionalLoadForFrameShared() and emit it through the BiDi event system,
providing clients with real-time navigation start notifications including URL, navigation ID,
and timing information.

The design reuses WebKit's existing automation session patterns, making it straightforward
to add the remaining navigation events (committed, failed, aborted, fragmentNavigated)
by hooking into their respective WebKit navigation callbacks. A note we needed a UInt128 for
Navigation so there is an added simple deterministic mapping to UInt64 used.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::navigationStartedForFrame):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):

Canonical link: https://commits.webkit.org/298524@main
